engine/runner.ml

```diff
@@ -711,7 +708,38 @@ let key_check () =
 
 let start_checks () = key_check ()
 
-let main repo_dir nocache deploy =
+let prune_nodes ns target_nodes =
+  let node_edges = List.map Node.get_edges ns |> List.concat in
+  let rev_edges = List.map (fun (x, y) -> (y, x)) node_edges in
+  let rec aux edges (targets : string list) =
+    let valid_edges =
+      List.filter
+        (fun (dst, _) -> List.exists (String.equal dst) targets)
+        edges
+    in
+    let new_targets =
+      List.fold_left
+        (fun ts (_, src) ->
+          match List.exists (String.equal src) ts with
+          | true ->
+              ts
+          | false ->
+              src :: ts)
+        targets valid_edges
+    in
+    if List.length targets == List.length new_targets then targets
```

Comment:
```suggestion
    if List.length targets = List.length new_targets then targets
```

Reply:
it's fine, one day I'll figure out which of these to use.

Reply:
We might get a bit of assistance - https://github.com/ocaml/ocaml/pull/9080